pkg/azure/rule_utils.go

```diff
@@ -0,0 +1,189 @@
+func CheckSecurityRulesCompliance(ctx context.Context, azureHandler *AzureSDKHandler, nsg *armnetwork.SecurityGroup) (bool, error) {
+	reservedPrioritiesInbound := make(map[int32]*armnetwork.SecurityRule)
+	reservedPrioritiesOutbound := make(map[int32]*armnetwork.SecurityRule)
+	err := setupMaps(reservedPrioritiesInbound, reservedPrioritiesOutbound, nil, nsg)
+	if err != nil {
+		utils.Log.Printf("An error occured during setup: %+v", err)
+		return false, err
+	}
+
+	// For Inbound Rules
+	priority, err := validateSecurityRulesConform(reservedPrioritiesInbound)
+	if err != nil {
+		if priority == -1 {
+			return false, fmt.Errorf("Non-compliant: %v", err)
+		}
+
+		_, err := setupAndCreateDenyAllRule(ctx, azureHandler, priority, inboundDirectionRule, *nsg.Name)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// For Outbound Rules
+	priority, err = validateSecurityRulesConform(reservedPrioritiesOutbound)
+	if err != nil {
+		if priority == -1 {
+			return false, fmt.Errorf("Non-compliant: %v", err)
+		}
+
+		_, err := setupAndCreateDenyAllRule(ctx, azureHandler, priority, outboundDirectionRule, *nsg.Name)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+// Checks that the NSG rules in a particular direction are conformant as per the description of (func) CheckSecurityRulesCompliance
+//
+// Returns the priority number of the deny all rule if it exists (with no error)
+// Or the priority number of the deny all rule to be created if the rules are non-conformant (with an error)
+// Or a priority number of -1 (with an error) to represent an invalid rule order.
+func validateSecurityRulesConform(reservedPriorities map[int32]*armnetwork.SecurityRule) (int32, error) {
+	var lowestRule *armnetwork.SecurityRule
+	lowestDenyPriorityNum := int32(maxPriority)
+	highestAllowPriorityNum := int32(minPriority)
+
+	for priority, rule := range reservedPriorities {
+		access := *rule.Properties.Access
+		if (access == armnetwork.SecurityRuleAccessAllow) && (priority > highestAllowPriorityNum) {
+			highestAllowPriorityNum = priority
+		}
+
+		if (access == armnetwork.SecurityRuleAccessDeny) && (priority <= lowestDenyPriorityNum) {
+			lowestRule = rule
+			lowestDenyPriorityNum = priority
+		}
+	}
+
+	// An allow rule's priority number is higher than a deny rule's priority number
+	if highestAllowPriorityNum > lowestDenyPriorityNum {
+		return -1, fmt.Errorf("Allow Rule with lower priority(%d) than Deny Rule(%d)", highestAllowPriorityNum, lowestDenyPriorityNum)
+	}
+
+	// No deny rule exists in the NSG, so non-conformant
+	if lowestRule == nil {
+		if reservedPriorities[maxPriority] != nil {
+			// The max priority number should not be associated to an allow rule
+			return -1, fmt.Errorf("Allow Rule at lowest priority(%d). Must be a deny all rule", maxPriority)
+		}
+
+		return maxPriority, fmt.Errorf("No deny rule present")
+	}
+
+	// If not a deny all rule, return priority to create a deny all rule
```

Comment:
Either way, we will need the new deny-all to override the default azure rules which are usually of the form:
{Priority: 65500, Name: Allow-all-vnet}
{Priority: 65501, Name: deny-all}
So no matter what, we want to put a deny-all rule with higher (lower-value) priority than these rules.